pymemcache/client/hash.py

```diff
@@ -0,0 +1,282 @@
+            self.add_server(server, port)
+
+    def add_server(self, server, port):
+        key = '%s:%s' % (server, port)
+
+        if self.use_pooling:
+            client = PooledClient(
+                (server, port),
+                **self.default_kwargs
+            )
+        else:
+            client = Client((server, port))
+
+        self.clients[key] = client
+        self.hasher.add_node(key)
+
+    def remove_server(self, server, port):
+        dead_time = time.time()
+        self._failed_clients.pop((server, port))
+        self._dead_clients[(server, port)] = dead_time
+        key = '%s:%s' % (server, port)
+        self.hasher.remove_node(key)
+
+    def _get_client(self, key):
+        _check_key(key, self.key_prefix)
+        if len(self._dead_clients) > 0:
+            current_time = time.time()
+            ldc = self._last_dead_check_time
+            # we have dead clients and we have reached the
+            # timeout retry
+            if current_time - ldc > self.dead_timeout:
+                for server, dead_time in self._dead_clients.items():
+                    if current_time - dead_time > self.dead_timeout:
+                        print('bringing server back in rotation')
+                        self.add_server(*server)
+                        self._last_dead_check_time = current_time
+
+        server = self.hasher.get_node(key)
+        print('got server %s' % server)
+        client = self.clients[server]
+        return client
+
+    def _safely_run_func(self, client, func, *args, **kwargs):
+        try:
+            if client.server in self._failed_clients:
+                # This server is currently failing, lets check if it is in
+                # retry or marked as dead
+                failed_metadata = self._failed_clients[client.server]
+
+                # we haven't tried our max amount yet, if it has been enough
+                # time lets just retry using it
+                if failed_metadata['attempts'] < self.retry_attempts:
+                    failed_time = failed_metadata['failed_time']
+                    if time.time() - failed_time > self.retry_timeout:
+                        print(failed_metadata)
+                        print('retrying')
+                        result = func(*args, **kwargs)
+                        # we were successful, lets remove it from the failed
+                        # clients
+                        self._failed_clients.pop(client.server)
+                        return result
+                    return
+                else:
+                    # We've reached our max retry attempts, we need to mark
+                    # the sever as dead
+                    print('marking as dead')
+                    self.remove_server(*client.server)
+
+            result = func(*args, **kwargs)
+            return result
+
+        # Connecting to the server fail, we should enter
+        # retry mode
+        except socket.error:
```

Comment:
Yeah, the hard part is thats kind of the whole point of this client.  You really do want a graceful handling of failures so that nodes can be added/removed as problems are detected.  I specifically only used `socket.error` for now so that I wasn't masking other exceptions (for example, if you pass larger than 1mb data a server error would still be raised).  We could add the `ignore_exc` flag back and have them pass that before this graceful behavior kicks in?

I'm not really a fan of the ignore_exc in the other clients, I think they should raise exceptions